crypto/evp/pkey_kdf.c

```diff
@@ -12,130 +12,187 @@
 #include <openssl/evp.h>
 #include <openssl/err.h>
 #include <openssl/kdf.h>
+#include <openssl/core.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
 #include "internal/numbers.h"
 #include "internal/evp_int.h"
 
+#define MAX_PARAM   20
+
+typedef struct {
+    EVP_KDF_CTX *kctx;
+    /* TODO(3.0): come up with a better way to do this */
```

Comment:
@levitte, this (already large) PR seems to be heading downhill fast.  I've got all of the base calls working fine but the legacy wrappers are causing lots of problems.

The latest is that delaying the `_set_params` has resulted in stack allocated things having their address taken (via helper the many macros) and the calling function is returning before the pointers are dereferenced.  I've worked around them here in a crude manner but a better solution should be possible.  I can quickly think of:

1. add OSSL_PARAM_allocate_... calls that copy their arguments into temporary malloced buffers;
2. expanding the param build calls to keep local copies of values; or
3. breaking strict backwards compatibility.

I'm sure there are other possibilities.


I suspect that fixing this will mean that an individual param will need to know if it was malloced or not.  This brings us back to a param _flags_ field.  We still haven't addressed the "have all of the params I sent been used" concern and the _flags_ field would be capable of this as well.


Also welcoming input from @mattcaswell and @slontis

Reply:
Hmmm, I see you do book keeping of what's allocated (palloc), so the need for a flag appears already taken care of.

As for the rest, I will have a closer look this evening. Can you have patience until then?

Reply:
I've a lot of patience here.  We need to determine a _good_ solution moving forwards -- we'll be stuck with it for a long time.  I do not believe that what I've done here is a good solution.

Fortunately, we've still got sufficient time to work out any wrinkles.  I think the infrastructure we've got in place is close but it needs some refinements to avoid this kind of nastiness.

For the moment, the question is about allocating temporary storage for things that could disappear (either from the stack or being freed).

Reply:
Let's also remember that this is bridge code. I hope it's going to be relatively short lived

Reply:
There is also a concern that some of the string argument will be stack allocated and disappear.  Our tests do not do this, but a user could.

I imagine that code like this exists in the wild:
```
int set_salt(EVP_PKEY_CTX *pctx, const char salt)
{
    char buf[];

    sprintf(buf, "salt is %s", salt);
    return EVP_PKEY_CTX_set1_scrypt_salt(pctx, buf, strlen(buf));
}
```
I.e. a suite of `OSSL_PARAM_allocate_XXX()` calls might be beneficial.  The param builder could alternatively be modified to keep copies of strings (it keeps copies of scalar values already) and have an `OSSL_PARAM_build_from_text()` helper function too.

Using `OSSL_PARAM_allocate_from_text()` works because it does copy the value unconditionally.  This presents a third option where the ctrl call (not the ctrl_str one) could convert to a string and then call `OSSL_PARAM_allocate_from_text()`.  A bit ugly and slow (which shouldn't matter) but it might be the cleanest of the solutions.

Reply:
So, the main reason this PR went down that particular rabbit hole is because you wanted to avoid multiple set_ctx_params calls, right?  This is quite a rad move, considering that the old ctrl functions basically just passed values to wherever they were immediately used.  I assume this is for speed reasons (although I'm not sure you're saving *that* much on speed, all things considered), and now you're suggesting something that, by your own account, is "a bit ugly and slow".

I'm starting to think that the cleanest path is really to go back to having the ctrl function simply call the set_ctx_params function, one ctrl at a time instead of caching like you currently do.

Reply:
Performance has nothing to do with this.  If the OSSL_PARAM code paths are ever performance critical, we've screwed up badly.

The real reason I went down the delayed all in one batching is because some of the old calls had side effects hardwired in.  The most prominent one was the set secret _string control_ silently adding a second control to clear the seed buffer.  Subsequent seed setting controls (either style) would append to the buffer.  To quote from one of my all time favourite man pages, this is best described as _unique and arcane_.

I don't like side effects like this and I think it best if they are permanently removed from the **_new_** APIs.  We are making an effort to improve things and persisting evil like this goes against that.

Thus far, all the string passing controls seem to be doing the right thing but we're not sure what user code will do.

As an aside, yes I would like to see batched param operations globally.  We've got the capability (almost) and it could greatly simplify some of the initialisation code (I'm pointing directly at the AES GCM initialisation mess here).

Reply:
> Let's also remember that this is bridge code. I hope it's going to be relatively short lived

I'm wondering what _short lived_ means here.  We can't deprecate it for years yet.  The policy is clear:
> No existing public interface can be removed until its replacement has been in place in an LTS stable release. The original interface must also have been documented as deprecated for at least 5 years. A public interface is any function, structure or macro declared in a public header file.

3.0 isn't LTS.

Reply:
That side effect isn't as silent as you think.  You should read https://github.com/openssl/openssl/blob/master/doc/man3/EVP_PKEY_CTX_set_tls1_prf_md.pod:

> EVP_PKEY_CTX_set_tls1_prf_secret() sets the secret value of the TLS PRF to seclen bytes of the buffer sec. Any existing secret value is replaced and any seed is reset.

Reply:
Which doesn't make it any less ugly :)

Reply:
What you *could* do is collect the seed in `pkey_kdf.c`, and only pass that down to the provider implementation when it's time to derive, and thereby support the resetting behaviour in `pkey_kdf.c`.  *The rest of the controls* can simply be passed down.  In other words, there's no reason you need to cache all parameters here just because you need to do so with the seeding material.

Reply:
This way, you can also avoid collecting a series of OSSL_KDF_PARAM_SEED in the provider implementation...  and we will have to modify libssl accordingly.

Reply:
Cache one, might as well cache them all.  We will encounter similar issues elsewhere.

Not collecting SEED calls in the provider might be a bonus -- it removes the order dependency.  I'm pretty sure that there are other such dependencies.

Reply:
"... might as well..."

Yeah well, it lead you to a downward spiral. Perhaps a rethink is in order here? Again, this is legacy support, there's no reason to do an all-in solution, especially when it ends up making your life harder. I do believe that the old KISS principle applies here.

Reply:
The point is that if I support the legacy code here the easy way, the APIs will never improve.
It is worth enduring some pain to make things better longer term.

Reply:
I have no idea what APIs you're talking about, and have no idea how this particular pile of legacy stuff would stop improvement of other APIs.

Reply:
The points are the mysterious side effects when setting a param (set the secret and clear the seed) and accumulating some of the params together.  Both could be avoided in the param layer by moving the special casing into the legacy layers (PKEY, direct calls) and into the various tests.  I've kind of half done this here.

Or are you suggesting that backward compatibility should be broken here?  And our such usages fixed (tests and TLS code).  If so, we'd risk breaking HKDF (concatenates params) and TLS1-PRF (concatenates params and has the side effect clearing).

Reply:
[TLS1-PRF](https://tools.ietf.org/html/rfc5246#section-5) *per se* doesn't concatenate bits of the seed as far as I can read the specs.  The EVP_PKEY implementation does so.
[HKDF](https://tools.ietf.org/html/rfc5869) *per se* doesn't concatenate bits of the info as far as I can read the specs.  The EVP_PKEY implementation does (it doesn't do the reset side-effect that the TLS1-PRF implementation does, though)

What I'm implying is that it's not *necessary* for the EVP_KDF implementation to support concatenation of info / seed bits, unless you absolutely want to...

Reply:
Another idea could be to support two params in TLS1-PRF, "seed", which just sets the see to the given value, and "seed++" or "add-seed", which appends seeding material to what's already there.

This does mean the EVP_MAC would have slightly changed semantics compared to the EVP_PKEY implementation, insofar that a seed reset will have to be explicit instead of being a side effect, i.e. when a reset is desired, use "seed" with a new value (could be zero bytes).

Same thing with HKDF' "info"

With that kind of solution, there should be no need whatsoever for the EVP_PKEY bridge to cache anything along the way.

Reply:
I thinking of reworking the param build APIs to save copies of all of the arguments.  Currently, it saves native types but not referential types, this could be a prime source of problems going forwards.

Reply:
It could take me a while, other priorities are interfering :(

Reply:
Intuitively, that sounds like the over engineering path.  I can't stop you from trying, but had to say it...

Reply:
(essentially, that sounds like you want to turn an OSSL_PARAM array into a store, and I very strongly believe that's the wrong way to go)

Reply:
I'd like to making build OSSL_PARAM arrays less prone to stack frames disappearing underneath them -- both Shane and I have been bitten hard by this.

The arrays of ints added here are a result of a helper macro invoking a helper function that passed the address of its integral argument as the param -- naturally the value disappeared and only later was it dereferenced.  This took way too long to track down.